### PR TITLE
Add RH image to Dockerfile

### PR DESCRIPTION
**Pull request by sigalsax**
Add Red Hat Base image to secrets-provider-for-k8s

_DOD_
- [x] Add RH certified image
- [x] Update publish process
- [ ] add a section to CONTRIBUTING.md regarding RH image (this section will be added once we get approval from RH)
- [x] a CHANGELOG entry for this change

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -100,8 +100,5 @@ pipeline {
     success {
       cleanupAndNotify(currentBuild.currentResult)
     }
-    unsuccessful {
```
Comment (sigalsax) [inline Jenkinsfile:103]: TODO: This will be restored before merge. I deleted this so we won't receive notifications while I work

File: Dockerfile [hunk 2]
```diff
@@ -64,3 +64,43 @@ COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provi
 
 ENTRYPOINT [ "/usr/local/bin/secrets-provider"]
 
+# =================== MAIN CONTAINER (REDHAT) ===================
+FROM registry.access.redhat.com/rhel as secrets-provider-for-k8s-redhat
+MAINTAINER CyberArk Software, Ltd.
```
Comment (orenbm) [inline Dockerfile:69]: the rest of the file doesn't have the comma

File: Dockerfile [hunk 3]
```diff
@@ -64,3 +64,43 @@ COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provi
...
+
+ARG VERSION
+
+LABEL name="secrets-provider-for-k8s"
+LABEL vendor="CyberArk"
+LABEL version="$VERSION"
+LABEL release="$VERSION"
+LABEL summary="Store secrets in Conjur or DAP and consume them in your Openshift application containers"
+LABEL description="The Secrets Provider for Kubernetes runs as an init container and fetches \
```
Comment (orenbm) [inline Dockerfile:78]: The CyberArk Secrets Provider for Kubernetes runs as an init container and populates the Kubernetes secrets required by the pod, from Conjur.

This is from the doc. If you think it's not good please change it also there.

File: Dockerfile [hunk 4]
```diff
@@ -64,3 +64,43 @@ COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provi
...
+                           /run/conjur && \
+    # We need open group permissions in these directories since OpenShift won't
+    # match our UID when we try to write files to them
+    chmod 770 /etc/conjur/ssl \
+              /run/conjur
+
+COPY LICENSE.md /licenses
+
+USER secrets-provider
+
```
Comment (orenbm) [inline Dockerfile:104]: missing this very important line: `COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provider /usr/local/bin/`

File: bin/build [hunk 5]
```diff
@@ -17,4 +18,11 @@ docker build \
...
 
-echo "---"
+echo "Building secrets-provider-for-k8s-redhat:$FULL_VERSION_TAG Docker image"
+
+docker build --tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" \
+   --build-arg VERSION="$FULL_VERSION_TAG" \
+   --target "secrets-provider-for-k8s-redhat" \
+   .
+
+echo "---"
```
Comment (orenbm) [inline bin/build:28]: missing new line

File: bin/build [hunk 6]
```diff
@@ -4,6 +4,7 @@ set -e
 
 . bin/build_utils
 
+TAG="secrets-provider-for-k8s"
```
Comment (orenbm) [inline bin/build:7]: why did you add this? i don't see it is used

File: Dockerfile (hunk 4 above)
Comment (orenbm) [reply Dockerfile:104]: I also see that in the authn-client we have also `ARG VERSION` - https://github.com/cyberark/conjur-authn-k8s-client/blob/master/Dockerfile#L116

I think that it's not necessary. can you please validate this and remove it from there if it is?

File: secrets.yml [hunk 7]
```diff
@@ -0,0 +1 @@
+REDHAT_API_KEY: !var redhat/projects/secrets-provider-for-k8s/api-key
```
Comment (orenbm) [inline secrets.yml:1]: missing new line

File: Dockerfile [hunk 8]
```diff
@@ -64,3 +64,43 @@ COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provi
...
+FROM registry.access.redhat.com/rhel as secrets-provider-for-k8s-redhat
+MAINTAINER CyberArk Software, Ltd.
+
+ARG VERSION
+
+LABEL name="secrets-provider-for-k8s"
+LABEL vendor="CyberArk"
+LABEL version="$VERSION"
+LABEL release="$VERSION"
+LABEL summary="Store secrets in Conjur or DAP and consume them in your Openshift application containers"
```
Comment (InbalZilberman) [inline Dockerfile:77]: I would add cyberark and maybe link to the docs?

File: bin/publish [hunk 9]
```diff
@@ -27,4 +28,19 @@ if [ "$GIT_DESCRIPTION" = "v${VERSION}" ]; then
...
+  docker tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" "${REDHAT_IMAGE}:${VERSION}"
+
+  if docker login scan.connect.redhat.com -u unused -p "${REDHAT_API_KEY}"; then
+    # you can't push the same tag twice to redhat registry, so ignore errors
+    if ! docker push "${REDHAT_IMAGE}:${VERSION}"; then
+      echo 'RedHat push FAILED! (maybe the image was pushed already?)'
+      exit 0
+    fi
+  else
+    echo 'Failed to log in to scan.connect.redhat.com'
```
Comment (InbalZilberman) [inline bin/publish:43]: @sigalsax will this fail the build? how would we know that the published failed?

File: Dockerfile (hunk 3 above)
Comment (sigalsax) [reply Dockerfile:78]: updating README

File: Dockerfile (hunk 4 above)
Comment (sigalsax) [reply Dockerfile:104]: VERSION comes as a build-arg from `./bin/build` which is the full tag

```
docker build --tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" \
   --build-arg VERSION="$FULL_VERSION_TAG" \
   --target "secrets-provider-for-k8s-redhat" \
   .
```

According to our [RH complaince docs](https://github.com/conjurinc/docs/blob/master/reference/infrastructure/openshift_registry_upload_guide.md), RH needs the following
```
LABEL version="$VERSION"
LABEL release="$VERSION"
```

Comment (orenbm) [reply Dockerfile:104]: i know that. the question is whether the `ARG VERSION` line is required. If the answer is no then we should remove it also from the authn-client to remove confusion.

File: bin/publish (hunk 9 above)
Comment (sigalsax) [reply bin/publish:43]: correct. I forgot the `exit 1`

File: Dockerfile (hunk 8 above)
Comment (sigalsax) [reply Dockerfile:77]: which docs, oss or dap?

File: Dockerfile (hunk 4 above)
Comment (sigalsax) [reply Dockerfile:106]: yes we need otherwise we get a warning during build that a variable was not given a value

Comment (sigalsax) [reply Dockerfile:106]: `[Warning] One or more build-args [VERSION] were not consumed`

File: CHANGELOG.md [hunk 10]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+- Added a RedHat-certified version of the image [cyberark/secrets-provider-for-k8s](https://github.com/cyberark/secrets-provider-for-k8s/pull/93)
```
Comment (orenbm) [inline CHANGELOG.md:10]: the word `Added` is redundant

Comment (orenbm) [reply CHANGELOG.md:10]: also you are missing the issue number in the link. It should be `cyberark/secrets-provider-for-k8s#93`

File: Dockerfile (hunk 8 above)
Comment (orenbm) [reply Dockerfile:77]: let's not add DAP docs in an open-source project.

File: Dockerfile (hunk 3 above)
Comment (orenbm) [reply Dockerfile:78]: can you update this as well?

File: Jenkinsfile [hunk 11]
```diff
@@ -98,7 +98,7 @@ pipeline {
 
   post {
     success {
-      cleanupAndNotify(currentBuild.currentResult)
+      cleanupAndNotify(currentBuild.currentResult, "#development", "@secrets-provider-for-k8s-owners")
```
Comment (orenbm) [inline Jenkinsfile:101]: why did you add a notification in case of success?

File: README.md [hunk 12]
```diff
@@ -11,9 +11,8 @@
 # CyberArk Secrets Provider for Kubernetes
 
 The CyberArk Secrets Provider for Kubernetes enables you to use secrets stored and managed in the CyberArk Vault 
-using DAP or Conjur and consume them in your Openshift application containers. To do so, the CyberArk Secrets 
-Provider for Kubernetes image runs as an init container and provides the Conjur secrets, required by the pod, 
-from DAP or Conjur.
+using DAP or Conjur and consume them in your Openshift application containers. To do so, the Secrets Provider for
```
Comment (orenbm) [inline README.md:14]: why did you remove the word `CyberArk`?

Comment (orenbm) [reply README.md:14]: please get a TW approval on this change (but why is it part of this PR?)

**Comment by orenbm** [comment]
@sigalsax please fix your commits:
image

File: Dockerfile (hunk 4 above)
Comment (sigalsax) [reply Dockerfile:106]: also when pushed to Openshift, the scan fails on lack of version

File: Jenkinsfile (hunk 11 above)
Comment (sigalsax) [reply Jenkinsfile:101]: I rebased off of master here. @eladkug ?

File: README.md (hunk 12 above)
Comment (sigalsax) [reply README.md:14]: You previously commented on the discrepancy between what was in the README and what I wrote in the description and from what I understood you said if I don't like it then change it and so I did. I can make it into a separate PR if that would be better and get TW approval. WDYT?

File: Jenkinsfile (hunk 11 above)
Comment (eladkug) [reply Jenkinsfile:101]: I didnt change the success section. 
I updated unsuccessful section only.

Comment (sigalsax) [reply Jenkinsfile:101]: weird.. okay I will just remove that again then

**Comment by sigalsax** [comment]
> @sigalsax please fix your commits:
> image

what's wrong with them? They explain what I added and I separated them out in terms of content writing and implementation of the task

File: Dockerfile (hunk 8 above)
Comment (sigalsax) [reply Dockerfile:77]: Let's solidify terminology and the type experience we want to offer to our customers. The github repo is named `Secrets Provider for K8s` but after a conversation w/ @InbalZilberman yesterday, we are naming the RH project `Secrets Provider for Openshift`. It is important to note that we also see traces of this shift in terminology for our authn client `Openshift Authenticator Client` although in our documentation we reference it as `Kubernetes Authenticator Client`.

Is this the direction we also want to go in? If so, should we remove traces of "Kubernetes" in our RH descriptions. For example, in the description, should it be:

```
The Secrets Provider for Openshift (for Kubernetes / for Kubernetes and Openshift) runs as an init container and fetches
```
I would like to get opinions here... @InbalZilberman @Tovli @orenbm

File: README.md [hunk 13]
```diff
@@ -11,9 +11,8 @@
 # CyberArk Secrets Provider for Kubernetes
 
 The CyberArk Secrets Provider for Kubernetes enables you to use secrets stored and managed in the CyberArk Vault 
-using DAP or Conjur and consume them in your Openshift application containers. To do so, the CyberArk Secrets 
-Provider for Kubernetes image runs as an init container and provides the Conjur secrets, required by the pod, 
-from DAP or Conjur.
+using DAP or Conjur and consume them in your Openshift application containers. To do so, the CyberArk Secrets Provider for
```
Comment (sigalsax) [inline README.md:14]: Should this be 
"using DAP or Conjur and consume them in your Kubernetes / Openshift application containers"? 
@orenbm

File: CHANGELOG.md [hunk 14]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+- RedHat-certified version of the secrets-provider-for-k8s image [cyberark/secrets-provider-for-k8s#93](https://github.com/cyberark/secrets-provider-for-k8s/pull/93)
```
Comment (orenbm) [inline CHANGELOG.md:10]: i think it should be `([cyberark/secrets-provider-for-k8s#93](https://github.com/cyberark/secrets-provider-for-k8s/pull/93))`

File: Dockerfile (hunk 8 above)
Comment (orenbm) [reply Dockerfile:77]: i don't have a strong opinion on this. I would like to get @izgeri 's opinion on this as she handled RH for the authn-client.

I'll go with whatever is decided in this thread.

File: README.md (hunk 13 above)
Comment (orenbm) [reply README.md:14]: i would use `Kubernetes` but will go with whatever @InbalZilberman decides.

Comment (InbalZilberman) [reply README.md:14]: Let's have both. Kubernetes / Openshift application

**Comment by orenbm** [comment]
> > @sigalsax please fix your commits:
> > image
> 
> what's wrong with them? They explain what I added and I separated them out in terms of content writing and implementation of the task

Their titles are bad. The title should state what changed or the outcome, and should not span more than 80 characters. The title "This commit adds the following:" is awful.

File: CHANGELOG.md (hunk 14 above)
Comment (sigalsax) [reply CHANGELOG.md:10]: https://github.com/cyberark/conjur-authn-k8s-client/edit/master/CHANGELOG.md looking at other projects I see we do a combo. ([]()) is better?

Comment (orenbm) [reply CHANGELOG.md:10]: yes, that's our standard

Comment (sigalsax) [reply CHANGELOG.md:10]: so i will update our whole changelog in that fashion. separate PR?

Comment (orenbm) [reply CHANGELOG.md:10]: only if it's easy, thanks

**Comment by sigalsax** [comment]
Blocked by RH certification approval and @boazmichaely said he will take the task from here in terms of configuring the portal and I will check back next week to see if we get confirmation from RH

File: CHANGELOG.md (hunk 14 above)
Comment (orenbm) [inline CHANGELOG.md:10]: please change `RedHat` to `Red Hat`

Comment (orenbm) [reply CHANGELOG.md:10]: also - this: https://github.com/cyberark/secrets-provider-for-k8s/pull/93/files#r430225230

File: bin/publish [hunk 15]
```diff
@@ -27,4 +28,19 @@ if [ "$GIT_DESCRIPTION" = "v${VERSION}" ]; then
...
   done
+
+  # Publish only latest to Redhat Registries
+  echo "Tagging and pushing ${REDHAT_IMAGE}"
+  docker tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" "${REDHAT_IMAGE}:${VERSION}"
+
+  if docker login scan.connect.redhat.com -u unused -p "${REDHAT_API_KEY}"; then
+    # you can't push the same tag twice to redhat registry, so ignore errors
+    if ! docker push "${REDHAT_IMAGE}:${VERSION}"; then
+      echo 'RedHat push FAILED! (maybe the image was pushed already?)'
```
Comment (orenbm) [inline bin/publish:39]: Red Hat

File: CHANGELOG.md [hunk 16]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+- Red Hat-certified version of the secrets-provider-for-k8s image ([cyberark/secrets-provider-for-k8s#93](https://github.com/cyberark/secrets-provider-for-k8s/pull/93))
```
Comment (orenbm) [inline CHANGELOG.md:10]: can we remove the hyphen? `Red Hat certified`

File: README.md [hunk 17]
```diff
@@ -13,10 +13,12 @@
...
-The CyberArk Secrets Provider for Kubernetes enables you to use secrets stored and managed in the CyberArk Vault 
-using DAP or Conjur and consume them in your Openshift application containers. To do so, the CyberArk Secrets 
-Provider for Kubernetes image runs as an init container and provides the Conjur secrets, required by the pod, 
-from DAP or Conjur.
+The CyberArk Secrets Provider for Kubernetes enables DAP to retrieve secrets stored and managed in the CyberArk Vault. The
+ secrets can be consumed by your Kubernetes or Openshift application containers. To retrieve the secrets from Conjur or DAP, 
+ the CyberArk Secrets Provider for Kubernetes runs as an init container and fetches the secrets that the pods require. 
+ 
+ To do so, the CyberArk Secrets Provider for 
+Kubernetes runs as an init container and fetches secrets that pods require from DAP or Conjur.
```
Comment (orenbm) [inline README.md:21]: this text has some duplications. Did it go by a TW?

Comment (sigalsax) [reply README.md:21]: that was removed in the previous force push ^

File: README.md [hunk 18]
```diff
@@ -13,11 +13,10 @@
...
 # CyberArk Secrets Provider for Kubernetes
 
-The CyberArk Secrets Provider for Kubernetes enables you to use secrets stored and managed in the CyberArk Vault 
-using DAP or Conjur and consume them in your Openshift application containers. To do so, the CyberArk Secrets 
-Provider for Kubernetes image runs as an init container and provides the Conjur secrets, required by the pod, 
-from DAP or Conjur.
-
+The CyberArk Secrets Provider for Kubernetes enables DAP to retrieve secrets stored and managed in the CyberArk Vault. The
+ secrets can be consumed by your Kubernetes or Openshift application containers. To retrieve the secrets from Conjur or DAP, 
+ the CyberArk Secrets Provider for Kubernetes runs as an init container and fetches the secrets that the pods require.
```
Comment (sigalsax) [inline README.md:18]: @orenbm yes Shuli reviewed this

File: README.md [hunk 19]
```diff
@@ -34,7 +33,7 @@ When we release a version, we push the following images to Dockerhub:
 1. Major.Minor
 1. Major
 
-In addition, each time the `master` build is green, we push a `<version>-<git_version>` (i.e. `0.2.0-d9494c1`) image to our internal repo.
+We also push the Major.Minor.Build image to our Red Hat registry.
```
Comment (sigalsax) [inline README.md:36]: @orenbm I also took the opportunity to remove the above info and say we also post the Major.Minor.Build version to the RH registry

**Comment by sigalsax** [comment]
UPDATE: after conversation with @InbalZilberman I am removing all public mentions of supporting a RH image. 

I will add the following back in once we get approval from @boazmichaely and RH on the image

in Contributing:
```
### Publish the git release
1. In the GitHub UI, create a release from the new tag and copy the change log
   for the new version into the GitHub release description.
1. The Jenkins pipeline auto-publishes new images to DockerHub, but to publish the Red Hat certified image you will need to visit its [management page](https://connect.redhat.com/project/4381831/view) and manually publish the image.
```

in README
```
We also push the Major.Minor.Build image to our Red Hat registry.
```

It is important to note that in the RH portal, auto publish is turned off so the work done in this PR  (including in the publish.sh) is non-damaging and not forward facing unless we switch the image forward facing.

File: README.md (hunk 19 above)
Comment (sigalsax) [reply README.md]: removed. this will be added back in once we get RH approval
